Modify lexer to allow single letter currencies (patch from #192)

cgevans opened the pull request:
I still need to check that this works after rebasing, and add tests, which is why I am marking it as a draft.

cgevans committed:
apply issue #192 patch to new lexer code

dnicolodi commented on beancount/parser/lexer.l:
```diff
@@ -218,6 +222,10 @@ ACCOUNTNAME     ([A-Z0-9]|{UTF-8-ONLY})([A-Za-z0-9\-]|{UTF-8-ONLY})*
 
+<AFTERDATE>[ \t\r]+ {
+    BEGIN(AFTERDATE);
+}
```
This looks funny and it seems to be required only because the `BEGIN(INITIAL)` added to `YY_USER_ACTION`. I think that getting rid of of both these would make the thing more understandable. I think the `BEGIN(INITIAL)` in `YY_USER_ACTION` can be placed in a default rule for the `<AFTERDATE>` state instead.

dnicolodi commented on beancount/parser/lexer.l:
```diff
@@ -166,6 +166,7 @@ int pyfile_read_into(PyObject *file, char *buf, size_t max_size);
         yylloc->last_column = yycolumn + yyleng - 1;                    \
         yylloc->file_name = yy_filename;                                \
         yycolumn += yyleng;                                             \
+        if (YY_START == AFTERDATE) BEGIN(INITIAL);                      \
```
The plan is to move the lexer to RE/flex where `YY_USER_ACTION` should go away. I would avoid to add more to it. Also, having a state transition in it does not make the mechanism really obvious. See below.

dnicolodi commented on beancount/parser/lexer.l:
```diff
@@ -301,15 +309,13 @@ ACCOUNTNAME     ([A-Z0-9]|{UTF-8-ONLY})([A-Za-z0-9\-]|{UTF-8-ONLY})*
+^[!&#?%PSTCURM]    {
+    yy_skip_line();
+    return SKIPPED;
 }
```
I like this. But I think `*` should be added to the flag characters. Also, it requires adjusting the grammar too: in a few places looks for an `HASH` or `ASTERISK` token (see the `txn` grammar rule).

cgevans replied:
Looking into this, `*`, `#`, and `:` were all handled separately before this line, with their own beginning-of-line checks. In fact, it's not clear to me that the previous code *ever* triggered for `#`.  Adding all three to this arrangement seems like it would be fine: I don't think there are places where any of them are used after a DATE, or after an INDENT, to mean anything other than a FLAG.  Then rules for separate HASH, ASTERISK, and COLON tokens can be moved to after this handling, with the separate beginning-of-line check removed.

However, I'm a bit uncomfortable here, because looking through the grammar, it appears that HASH is actually used in what appears to be an undocumented but implemented compound cost statement, and ASTERISK is used in an undocumented and apparently not yet implemented cost merging statement.  While I think my new changes preserve these behaviours, I'm hoping that they don't break any other potential behaviour I'm not aware of.

cgevans replied:
Hmm... there seems to be a wider problem here.  `beancount.core.flags` has `[*!PSTCURM]` as valid flags, as I now notice you pointed out in #516.

blais replied:
"#" is used in compound costs, to specify unit and total cost, as {unit # total currency}.

cgevans replied:
Yes.  Before this change, "#" was tokenized early on as HASH if not the first token, and otherwise the line was skipped.  "#" does not appear to ever trigger line 304, as it is caught by line 257.  In the grammar, it can be parsed as a flag, but that doesn't fit with beancount.core.flags (see also https://github.com/beancount/beancount/issues/516).  

After the change, "#" is never parsed as a flag, per beancount.core.flags.  The line-skipping and "#" in a compound cost is not affected by the change.  Alternatively, "#" could be added to line 275 to be parsed as a flag, without affecting either of hose.

dnicolodi reviewed (commented):
How is this handling flags that can be prepended to postings? Like
```
2020-07-28 * "Test"
  Expenses:Foo  1.00 USD
  C Assets:Cash
```

cgevans commented on beancount/parser/lexer.l:
```diff
@@ -382,7 +389,7 @@ NULL		{
 
  /* Currencies. These are defined as uppercase only in order to disambiguate the
   * syntax. This is kept in sync with beancount.core.amount.CURRENCY_RE. */
-[A-Z][A-Z0-9\'\.\_\-]{0,22}[A-Z0-9]	{
```
I actually need to add one to the maximum length here in order to preserve the 24 character limit that is noted in the documentation.  But is there a reason for that limit?

cgevans committed:
Simplify flag lex+gramm, sync flags, fix posting

These changes use a default rule instead of user action to switch back
from the possible flag start condition to INITIAL.  The possible flag
condition has been changed to
also occur after an indent, to
recognize posting flags, and renamed
accordingly to POSSIBLEFLAG.

Asterisk, colon, and hash lexing has been simplified: they are now
directly recognized as flags in the lexer when they are flags, removing
the need for checks in the grammar.

The lexer has been changed to only allow flags that are in
beancount.core.flags, as described in the comments and language syntax.
It has also been changed to no longer allow
arbitrary flags at the beginning of the line
as a way to skip that line.
(https://github.com/beancount/beancount/issues/516)

Note that these last two changes might break files that currently work,
but don't actually comply with the language syntax documentation.

cgevans committed:
Fix regression on 24 char limit for currencies

cgevans commented:
> How is this handling flags that can be prepended to postings? Like
> 
> ```
> 2020-07-28 * "Test"
>   Expenses:Foo  1.00 USD
>   C Assets:Cash
> ```

It doesn't, because I completely forgot about that.  The commit fixes this.  However, I need to have Bison 3.6 installed to test the grammar changes, which will take some time.

blais commented:
A bit reluctant to make this change in v2 and break things (unless I know you'll be there to support it); I think it may be wise to do this in v3 along with the RE/flex conversion.

blais commented:
We want to fix this more generally in v3, with a scanner rewrite. I'm really reluctant to change the v2 lexer's state at this moment.
Let's hold off this feature for v3.

cgevans commented:
@blais, while this could be fixed more generally with a rewrite, it seems from the mailing list that v3 is in planning stages, involves considerable changes, and has a relatively long timescale.  This change is relatively small, and while it currently also addresses https://github.com/beancount/beancount/issues/516, could be changed to not address this, and would then involve minimal changes: in that case, the only change would be that a single capital letter would only be interpreted as a flag if it came after a date or indent, while flag and first-token behaviour would not change.  Would this smaller change be better?
